### PR TITLE
use vcpkg for Windows & macOS builds; add Qt6 builds; use sccache

### LINKED ISSUE #78: Add CI for using macOS with clang
Github Actions supports using macOS as the machine, add a CI definition which supports this ( https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources ).

- [ ] clang
- [ ] Rust (cargo, clippy, rustfmt, rustc) (also mdbook and mdbook serve eventually)
- [ ] Qt 5.15 / Qt 6.2
- [ ] clang-format
- [ ] cmake
- [ ] reuse
- [ ] valgrind

### PR DESCRIPTION
**Pull request by Be-ing**
This builds Qt with vcpkg and caches the C++ dependencies on GitHub Packages. The builds work on Linux, Windows, and macOS for both Qt5 and Qt6, however the tests only pass on Linux. It uses sccache for both C++ and Rust to speed up the builds.

Currently it is using GitHub Packages under my GitHub user because that is what I have write access to. GITHUB_TOKEN can only have write access to the KDAB organization's GitHub Packages for a pull request from a branch on the upstream repository; it only has read access for pull requests from forked repositories. When this is merged, `Be-ing` should be replaced with `KDAB` in the `Authenticate to GitHub Packages` step of the GitHub Actions workflow and a GitHub Personal Access Token with read-only permission to packages generated for anonymous access in CMakeLists.txt.

Fixes #78, #79

### CODE REVIEW

**Comment by Be-ing** [comment]
qmltestrunner fails to import the QtQuick module on macOS. I suspect this has something to do with statically linking Qt. I tried specifying the `-import` and `-plugins` command line options for qmltestrunner using the absolute path to where vcpkg installed Qt, but those did not work. For now I have disabled the one test (example_qml_extension_plugin_test) which uses that. Otherwise macOS tests are working now.

Windows Qt5 tests are failing to link symbols that look like they should come from `cxx-qt-gen/src/cxx_qt.cpp`. I don't know what's going wrong there.

Windows Qt6 tests are hanging indefinitely.

File: .github/workflows/github-cxx-qt-tests.yml
Comment (ahayzen-kdab) [inline .github/workflows/github-cxx-qt-tests.yml:38]: Note that #98 should fix the valgrind tests for Qt 6 on Linux :-)

File: cmake/CxxQt.cmake
Comment (ahayzen-kdab) [inline cmake/CxxQt.cmake:28]: Is this something we should report upstream to CXX ? or is it to do with how CMake is being used here?

File: cxx-qt-build/src/lib.rs
Comment (ahayzen-kdab) [inline cxx-qt-build/src/lib.rs:88]: maybe add a comment for this method describing why it's needed

**Comment by ahayzen-kdab** [comment]
> qmltestrunner fails to import the QtQuick module on macOS. I suspect this has something to do with statically linking Qt. I tried specifying the `-import` and `-plugins` command line options for qmltestrunner using the absolute path to where vcpkg installed Qt, but those did not work. For now I have disabled the one test (example_qml_extension_plugin_test) which uses that. Otherwise macOS tests are working now.
> 
> Windows tests are failing to link symbols that look like they should come from `cxx-qt-gen/src/cxx_qt.cpp`. I don't know what's going wrong there. Maybe not worth spending time on before #105 is merged.
> 
> Windows Qt6 tests are hanging indefinitely.

So the macOS part sounds interesting, which part are we static linking to Qt ? I thought we should be statically linking the Rust library into the C++ library or executable and then that is dynamic linking to Qt ? Or is something going wrong / missing on macOS ?

For Windows, as you've managed to get things building quite far, I'll try and setup an environment so I can investigate what is failing on Windows (as none of us have tried it yet :sweat_smile: ), and as you have noted parts there are potential issues in cxx_qt.cpp are being changed in #105 so it'll likely be after that.

File: cmake/CxxQt.cmake
Comment (Be-ing) [reply cmake/CxxQt.cmake:28]: I'm unsure. I tested cloning the cxx repository and building that locally on macOS and could not reproduce the issue, so I think it might have something to do with invoking cargo from CMake, but I don't know why that would make a difference.

Comment (Be-ing) [reply cmake/CxxQt.cmake:28]: >  fatal error: 'string.h' file not found

That was the error I was running into.

**Comment by Be-ing** [comment]
> So the macOS part sounds interesting, which part are we static linking to Qt ?

vcpkg defaults to building everything statically on macOS (as well as Linux, though we're not using vcpkg for Linux builds).

> Or is something going wrong / missing on macOS ?

I haven't tested static linking on another OS or dynamic linking on macOS, but that may be worth a try.

File: cxx-qt-build/src/lib.rs
Comment (Be-ing) [reply cxx-qt-build/src/lib.rs:88]: done in #108

**Comment by Be-ing** [comment]
> I'll try and setup an environment so I can investigate what is failing on Windows

Note that this branch will automatically download the dependencies with vcpkg when configuring CMake using my read-only PAT. This takes a while (several minutes) during which there isn't much feedback from vcpkg, but don't worry, it hasn't frozen.

File: cmake/CxxQt.cmake
Comment (Be-ing) [reply cmake/CxxQt.cmake:28]: I tried replacing this line with `set(CMAKE_CXX_STANDARD 17)` but it did not work. Curiously, when I manually specified `-D CMAKE_CXX_STANDARD=17` on the command line it worked. 🤔

Comment (Be-ing) [reply cmake/CxxQt.cmake:28]: Let's find out with the cxx maintainers think: https://github.com/dtolnay/cxx/issues/1020

**Comment by Be-ing** [comment]
> qmltestrunner fails to import the QtQuick module on macOS. I suspect this has something to do with statically linking Qt.

I suspect this may be a bug in qmltestrunner's build system rather than in cxx-qt.

**Comment by Be-ing** [comment]
How shall we proceed with this PR? I think it would be reasonable to disable the tests for Windows, merge this, then you could fix the Windows tests in a future PR.

File: .github/workflows/github-cxx-qt-tests.yml
Comment (Be-ing) [inline .github/workflows/github-cxx-qt-tests.yml:171]: Note that this TODO and the TODO in CMakeLists.txt for generating a PAT will need to be taken care of by you after merging this. I can't push vcpkg packages to KDAB's GitHub Packages because this is a PR from a forked repository so GITHUB_TOKEN does not have write permission.

You could merge this PR as is then take care of this afterwards. Alternatively, you could merge/cherry pick the commits besides the one introducing vcpkg, then cherry pick the vcpkg commit and address the TODOs in one modified commit.

**Comment by ahayzen-kdab** [comment]
> How shall we proceed with this PR? I think it would be reasonable to disable the tests for Windows, merge this, then you could fix the Windows tests in a future PR.
> 
> If you want to speed up the CI runs, manually cancel the earlier runs from https://github.com/KDAB/cxx-qt/actions. The earlier runs will take many hours before they finish because they didn't hit the GitHub Packages cache. I don't have permission to cancel CI runs on this repo.

I'll try and have a closer look at this tomorrow, I wonder about splitting things into smaller changes as we have multiple things going on here. eg there could be the following pull requests (?)

- Add Qt 6 support for Linux
- Add macOS support (Qt 5 and Qt 6)
- Add Windows support (Qt 5 and Qt 6) (we could then leave this one open until we figure out the changes to get the tests passing :-) ) (and probably stack this branch ontop of the macOS one, as the macOS one looks likely to land next)

**Comment by Be-ing** [comment]
> Add Qt 6 support for Linux

The rewrite of the GitHub Actions workflow file and adding vcpkg for macOS & Windows is all done together in one commit. If you insist I could split it up, but IMO that wouldn't be worth the trouble.

**Comment by Be-ing** [comment]
I disabled the Windows tests that fail to link and now CI passes in every condition (Linux, Windows, and macOS with both Qt5 and Qt6).

**Comment by ahayzen-kdab** [comment]
OK, i've merged #109 I'll take a look at this tomorrow :-) I'll need to flip the CI settings to require the individual checks rather than "ctest" it is currently.

And I'll need to read about how the GitHub packages works, is it possible for me to add a commit into this branch with the right tokens and change the url or will that not work as this is coming from a fork? As note that all contributions (even my own) come via a forked repository so how will they work ? :thinking:

**Comment by Be-ing** [comment]
> is it possible for me to add a commit into this branch with the right tokens and change the url or will that not work as this is coming from a fork?

That will not work because this branch is from a fork.

>  As note that all contributions (even my own) come via a forked repository so how will they work ?

Write access to GitHub Packages only matters when updating packages from vcpkg by pulling updates from https://github.com/microsoft/vcpkg into the vcpkg submodule. If GITHUB_TOKEN doesn't have write access on those PRs, CI will still work, but it will take a long time (Qt takes several hours to build on GitHub Actions) and the packages vcpkgs builds would be rebuilt when the PR is merged.

The read-only Personal Access Token is needed so that developers can simply run `cmake -S . -B build` on Windows and macOS and vcpkg will download the prebuilt packages from GitHub Packages instead of taking a few hours to build Qt locally.

**Comment by Be-ing** [comment]
CI jobs with caching are currently taking about 20 minutes. sccache statistics show that barely any of the C++ compilation is hitting the cache. I'm not sure why that is happening, but addressing that could probably save a few minutes.

**Comment by Be-ing** [comment]
> CI jobs with caching are currently taking about 20 minutes. sccache statistics show that barely any of the C++ compilation is hitting the cache.

I fixed this by recycling some CMake code from other projects as a CMake module. CI jobs now take ~16 minutes. A lot of that time is [restoring vcpkg packages from GitHub Packages](https://github.com/microsoft/vcpkg/discussions/23470).

**Comment by Be-ing** [comment]
> is it possible for me to add a commit into this branch with the right tokens and change the url or will that not work as this is coming from a fork?

You could push this branch to a branch on the upstream repository, make a new pull request, then add a commit changing the GitHub Packages credentials.

File: CMakeLists.txt
Comment (ahayzen-kdab) [inline CMakeLists.txt:13]: Please add another line with `SPDX-FileContributor: Be <be.0@gmx.com>` at the top :-)

File: cmake/CompilerCaching.cmake
Comment (ahayzen-kdab) [inline cmake/CompilerCaching.cmake:1]: Are you able to link to where this file came from? So i can check the license information :-)

File: vcpkg.json
Comment (ahayzen-kdab) [inline vcpkg.json:11]: On Linux we also have the following do we need these ? or are they included already?

- qml-module-qtquick2
- qml-module-qttest

**Review by ahayzen-kdab — CHANGES_REQUESTED** [review]
It's looking good, left some more comments for clarifications :-)

File: CMakeLists.txt
Comment (ahayzen-kdab) [reply CMakeLists.txt:16]: Ah, or is the problem that you haven't downloaded Qt yet, so it can't know what major version it is?

In that case this option is only relevant when VCPKG is enabled? so doesn't need to be set on the Linux runners and can this option be behind the if(VCPKG) statement or is that not possible?

Comment (ahayzen-kdab) [reply CMakeLists.txt:16]: Could potentially even be called VCPKG_ENABLE_QT6 or something to make it obvious that this isn't choosing to use Qt 6 for all platforms, but telling VCPKG which Qt versions to build and use?

File: .github/workflows/github-cxx-qt-tests.yml
Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:112]: actions/checkout doesn't support moving the path that the repository is cloned into unless it's under the default GITHUB_WORKSPACE. Unfortunately the default D drive runs out of space when vcpkg builds Qt on Windows, so this has to be moved to the C drive.

File: .github/workflows/github-cxx-qt-tests.yml
Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:129]: yes

File: cmake/CompilerCaching.cmake
Comment (Be-ing) [reply cmake/CompilerCaching.cmake:1]: It originally came from https://github.com/tenacityteam/tenacity and I split it out to a separate file in https://github.com/nextcloud/desktop/pull/3947

File: vcpkg.json
Comment (Be-ing) [reply vcpkg.json:11]: Evidently those are not needed considering the macOS tests pass without them.

File: .github/workflows/github-cxx-qt-tests.yml
Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:20]: It is built from source in the `Build Rust tools` step.

File: cmake/CompilerCaching.cmake
Comment (ahayzen-kdab) [reply cmake/CompilerCaching.cmake:1]: OK you could add you and other authors as `SPDX-FileContributor` lines if you wanted, and I see how you as the original author have relicensed it out to BSD etc thanks!

File: vcpkg.json
Comment (ahayzen-kdab) [reply vcpkg.json:11]: Right, just curious how there were linking issues to QtQuick at one point, wondered if it was related. Anyway as you say this works enough and we can continue tweaking in further changes.

File: .github/workflows/github-cxx-qt-tests.yml
Comment (ahayzen-kdab) [reply .github/workflows/github-cxx-qt-tests.yml:20]: Ah i see it comes from Rust / Cargo, i assumed it'd be coming from the Ubuntu repos. Cool ! :-)

File: CMakeLists.txt
Comment (Be-ing) [reply CMakeLists.txt:20]: > most of us

Are you referring to KDAB employees? I think you're the exception having Qt already installed for development on macOS & Windows. Do you already have it installed somewhere CMake knows to look for it?

File: CMakeLists.txt
Comment (Be-ing) [reply CMakeLists.txt:16]: > Ah, or is the problem that you haven't downloaded Qt yet, so it can't know what major version it is?

That plus picking which version of Qt to use when both Qt5 and Qt6 are installed. Does specifying QT_VERSION_MAJOR as an argument to CMake work for picking which version to use when both are installed and using version-agnostic CMake code?

File: CMakeLists.txt
Comment (ahayzen-kdab) [reply CMakeLists.txt:20]: Right, "most of us" as in KDAB employees or Qt developers. Maybe it's OK to leave it on, it just might not be obvious what's happening when it goes off and downloads a load of things when you already have Qt configured. I'm unsure of exactly the best opt-in vs opt-out here :-)

File: .github/workflows/github-cxx-qt-tests.yml
Comment (ahayzen-kdab) [reply .github/workflows/github-cxx-qt-tests.yml:112]: Unfortunate :-) Can you document this in a comment, so that we know why this is done and what would be required to move to the github maintained one later etc.

File: CMakeLists.txt
Comment (ahayzen-kdab) [reply CMakeLists.txt:16]: Good question, I use containers to isolate my environments as different projects need different versions of everything :-D But I believe others use things like `qtchooser` which uses the QT_SELECT env variable to pick ? Or you just change your PATH to point to the qmake file of the version of Qt you want.

Then reading https://doc.qt.io/qt-6/cmake-qt5-and-qt6-compatibility.html it suggests that the CMake option of `QT_DEFAULT_MAJOR_VERSION` is the way to do it when you have both and then the find_package() will set `QT_VERSION_MAJOR` to the corresponding value.

Comment (Be-ing) [reply CMakeLists.txt:17]: > QT_DEFAULT_MAJOR_VERSION

Oh, thanks! I'll use that variable to determine which Qt version to tell vcpkg to use.

Comment (ahayzen-kdab) [reply CMakeLists.txt:17]: If it isn't set, should we just default to Qt 6 i wonder ? Eg default to the "latest and greatest" :-)

Comment (ahayzen-kdab) [reply CMakeLists.txt:17]: (note this should only default for vcpkg itself and not on Linux where we want to to scan with find_package)

File: cmake/CompilerCaching.cmake
Comment (ahayzen-kdab) [inline cmake/CompilerCaching.cmake:1]: This should be

```
# SPDX-FileCopyrightText: 2021 Tenacity Audio Editor contributors
# SPDX-FileContributor: Be <be.0@gmx.com>
# SPDX-FileContributor: Emily Mabrey <emabrey@tenacityaudio.org>
#
# SPDX-License-Identifier: BSD-3-Clause
```

Comment (ahayzen-kdab) [reply cmake/CompilerCaching.cmake:1]: (to fix the `reuse` failure :-) )

File: CMakeLists.txt
Comment (Be-ing) [reply CMakeLists.txt:21]: Maybe that will work. Let's find out if `find_package` works before `project` :)

**Comment by Be-ing** [comment]
Well, this is strange. Explicitly setting `QT_DEFAULT_VERSION_MAJOR=6` with both Qt5 and Qt6 development packages installed on Fedora 35 fails with:

```
CMake Error at /usr/lib64/cmake/Qt5QmlImportScanner/Qt5QmlImportScannerConfig.cmake:11 (qt6_import_qml_plugins):
  Unknown CMake command "qt6_import_qml_plugins".
Call Stack (most recent call first):
  examples/qml_extension_plugin/CMakeLists.txt:44 (qt_import_qml_plugins)
```

`qt6_import_qml_plugins` is defined in `/usr/lib64/cmake/Qt6Qml/Qt6QmlMacros.cmake`. I am puzzled why `Qt5QmlImportScannerConfig.cmake` is invoked at all. Not setting `QT_DEFAULT_VERSION_MAJOR` works fine and defaults to Qt6 as expected. Setting `QT_DEFAULT_MAJOR_VERSION=5` also works. I suppose there's a bug in Qt's CMake scripts?? Anyway, it's working now.

**Comment by Be-ing** [comment]
I think all review comments have been addressed at this point. I also added documentation to README.md explaining how the build system works.

**Comment by ahayzen-kdab** [comment]
> Well, this is strange. Explicitly setting `QT_DEFAULT_VERSION_MAJOR=6` with both Qt5 and Qt6 development packages installed on Fedora 35 fails with:
> 
> ```
> CMake Error at /usr/lib64/cmake/Qt5QmlImportScanner/Qt5QmlImportScannerConfig.cmake:11 (qt6_import_qml_plugins):
>   Unknown CMake command "qt6_import_qml_plugins".
> Call Stack (most recent call first):
>   examples/qml_extension_plugin/CMakeLists.txt:44 (qt_import_qml_plugins)
> ```
> 
> `qt6_import_qml_plugins` is defined in `/usr/lib64/cmake/Qt6Qml/Qt6QmlMacros.cmake`. I am puzzled why `Qt5QmlImportScannerConfig.cmake` is invoked at all. Not setting `QT_DEFAULT_VERSION_MAJOR` works fine and defaults to Qt6 as expected. Setting `QT_DEFAULT_MAJOR_VERSION=5` also works. I suppose there's a bug in Qt's CMake scripts?? Anyway, it's working now.

Right sounds like a bug to report upstream :-)

File: cmake/CxxQt.cmake
Comment (ahayzen-kdab) [reply cmake/CxxQt.cmake:28]: Lets leave this upstream for now, I'll likely open an issue for each of the FIXME / TODOs from this branch when it lands.

File: .github/workflows/github-cxx-qt-tests.yml
Comment (ahayzen-kdab) [inline .github/workflows/github-cxx-qt-tests.yml:108]: If this only affects Windows, i assume it's not simple to use this only for Windows and use the normal checkout system for Linux and macOS?

Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:108]: That would be possible, but I'd like to minimize platform-specific code.

Comment (ahayzen-kdab) [reply .github/workflows/github-cxx-qt-tests.yml:108]: OK, do we know if there are any issues upstream to have this as a feature request in the normal checkout command?

Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:110]: Yes and I added a comment: https://github.com/actions/checkout/issues/197

Comment (ahayzen-kdab) [reply .github/workflows/github-cxx-qt-tests.yml:109]: Thanks! Perfect!

Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:109]: (To clarify, I added a comment to the workflow file in this repo, not commented on the issue.)

File: .github/workflows/github-cxx-qt-tests.yml
Comment (ahayzen-kdab) [reply .github/workflows/github-cxx-qt-tests.yml:162]: Ah I've found where to create personal access tokens, but for my account. I'll need one for the organisation...

Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:162]: > Generate a read-only token form somewhere (where do i do this? sweat_smile ) for the kdab packages?

Yes, do that here: https://github.com/settings/tokens
Be sure to not paste the entire string verbatim into CMakeLists.txt and read the comments there.

> The GITHUB_TOKEN in github-cxx-qt-tests I just leave and is it auto generated ? or do i need to add this as a "secret" into the settings?

Yes, GITHUB_TOKEN is generated automatically by GitHub Actions. You do not need to manually add a secret. https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow

> Then once I have all of that, i commit it and push directly into the main branch, and watch CI run on the commit and build the packages (will take a while).

You don't need to do it on the main branch. You could push to a branch on the upstream (KDB/cxx-qt) repo rather than your user's fork. And yes it will take like 3-4 hours.

> Then future pull requests should pick up the read-only one from CMake ?

Pull requests from forks and well as local builds will have read-only access. When you want to update dependencies, simply merge the master branch of https://github.com/microsoft/vcpkg into the vcpkg submodule and push that either to the main branch or another branch on the upstream repository.

Comment (ahayzen-kdab) [reply .github/workflows/github-cxx-qt-tests.yml:162]: Or it seems like you can't for orgs, but if your user account has permission for that org then you can use tokens you generate. So a token i create might be enough.

Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:162]: > Ah I've found where to create personal access tokens, but for my account. I'll need one for the organisation...

I don't think they exist at the organization level. Generating one for your user is fine. Make sure to set it to not expire automatically after a time limit when you create it.

Comment (ahayzen-kdab) [reply .github/workflows/github-cxx-qt-tests.yml:162]: Cool, I see thanks! Just double checking with the org owners that they are OK with this setup :-)

Comment (Be-ing) [reply .github/workflows/github-cxx-qt-tests.yml:162]: It may help for them to know that GitHub Packages is free for public repositories and there are no documented limitations on storage space or bandwidth: https://github.com/features/packages#pricing (look at the left side, not the table)

**Comment by ahayzen-kdab** [comment]
OK, lets go with this!

**Comment by ahayzen-kdab** [comment]
@Be-ing Thanks for your contribution! I wasn't expecting to have this working for a while! :-)

I'll try updating the tokens, I wonder if the "username" is my account or the org account if the token is from my personal account :thinking: ...

**Comment by Be-ing** [comment]
> I wonder if the "username" is my account or the org account

Yes, it's confusing, but the user for the GitHub Packages NuGet source is the organization, not ahayzen-kdab.

**Comment by ahayzen-kdab** [comment]
Thanks :-)

**Comment by Be-ing** [comment]
The [CI build](https://github.com/KDAB/cxx-qt/runs/5500106856) on the `update-tokens` branch succeeded. You can merge that to the main branch now.

> I wasn't expecting to have this working for a while! :-)

Yep, I've done this for two different cross platform C++ applications before, which took months of work, so I figured I'd help out with this cool project. In my experience this setup with vcpkg, GitHub Actions, and GitHub Packages has been the most practical way to maintain cross platform C++ applications. Don't hesitate to ping me if you have any questions maintaining it. The same build process works both on CI and for local development, so you don't get ugly surprises when developers try to get started contributing nor weird issues in CI builds that developers can't reproduce locally. Also, vcpkg works on Linux, so you can maintain packaging for Windows and macOS from Linux without needing to actually use Windows or macOS much.

P.S. I just sent an application to careers@kdab.com. It would be great to continue working together on this project. :)

**Comment by ahayzen-kdab** [comment]
Thanks! It's all in main now :-) I'll tag you if we have any further issues thanks! And good luck ! :-)

**Pull request merged**
